### ISSUE TITLE
random keys: fold_in and seed from string/bytes

### ISSUE THREAD

**Issue by ZacCranko** [google/jax#14239]
There is an issue with `jax.random.split` which is that the number of keys you might want will change as developing, e.g. you suddenly discover you need an extra one here or there. 

I thought maybe a good solution to this would be either adding `__attrgetter__` methods to "PRNGKeys" so you can do something like `key.some_string`  getting aliased to `jax.random.fold_in(key, hash(some_string))` or alternately add an overload to `fold_in` that accepts strings. This would lend itself to nice, self-documenting, patterns like

```python
sampler = jax.random.PRNGKey(sampler_seed)
initial_noise = jax.random.normal(fold_in(sampler, "initial_noise"), shape)
```

Soupey twist!

**Comment by froystig** [comment]
Thanks for filing! To what extent does the request here also include the solution? Does folding in `hash(message)` present any issue?

**Comment by jakevdp** [comment]
One issue with `hash(message)` is that by design, the hash is salted with a different value in each Python runtime. If we want results to be reproducible from run to run, we should use one of the deterministic hash functions from `hashlib` instead.

**Comment by froystig** [comment]
Good point. Consider my `hash` pseudocode, then (standing in for `hashlib` use or for anything else that's reasonable and deterministic).

**Comment by ZacCranko** [comment]
Thanks for tip @jakevdp. Is there a hash function you'd recommend?

**Comment by jakevdp** [comment]
You could do something like this with the md5 hash:
```python
In [1]: import hashlib

In [2]: my_hash = lambda x: int(hashlib.md5(x.encode()).hexdigest(), base=16)

In [3]: my_hash('hello world')
Out[3]: 125893641179230474042701625388361764291
```

**Comment by ZacCranko** [comment]
Ah, not quite as simple as a drop-in to compose with `fold_in`, but good to know. Much appreciated.

**Comment by froystig** [comment]
Overall, there's freedom in how to go from strings to ints (`hashlib` alone offers several choices), with no clear principle to favor one approach uniformly over another. I think it's reasonable for JAX to continue to offer `fold_in` as the basic primitive, and leave the rest up to users and libraries. Would you agree?

**Comment by ZacCranko** [comment]
I think having this functionality built into jax would be convenient, but I understand the issue of a lack of a canonical choice. 

By the way, if anyone comes across this issue, you will need to modify @jakevdp's snippet to truncate the 128 bit md5 output to 32bit:
```python
lambda x: int(hashlib.md5(x.encode()).hexdigest(), base=16) & 0xfffffff
```

**Comment by jakevdp** [comment]
You can also slice the `hexdigest` to a specified number of entries. Since each hexcode is 4 bits, 8 entries is 32 bits:
```python
lambda x: int(hashlib.md5(x.encode()).hexdigest()[:8], base=16)
```
